### PR TITLE
Properly convert and translate unit structs into MIR

### PR DESCRIPTION
**Pull request by nagisa**
Fixes https://github.com/rust-lang/rust/issues/30514

i also went ahead and removed the redundant `ItemKind`s.

r? @nikomatsakis (this is an easy one I guess)

### CODE REVIEW

**Comment by nagisa** [comment]
```rust
struct Droppable;
fn mir() {
    Droppable
}
```

results in

foo

**Comment by luqmana** [comment]
LGTM! @bors: r+

File: src/librustc_mir/hair/cx/expr.rs [hunk 1]
```diff
@@ -519,51 +519,51 @@ fn convert_path_expr<'a, 'tcx: 'a>(cx: &mut Cx<'a, 'tcx>, expr: &'tcx hir::Expr)
...
+        // A regular function.
+        def::DefFn(def_id, _) => (def_id, ItemKind::Function),
+        def::DefMethod(def_id) => (def_id, ItemKind::Method),
+        def::DefStruct(def_id) => {
+            match cx.tcx.node_id_to_type(expr.id).sty {
+                // A tuple-struct constructor.
+                ty::TyBareFn(..) => (def_id, ItemKind::Function),
+                // This is a special case: a unit struct which is used as a value. We return a
+                // completely different ExprKind here to account for this special case.
+                ty::TyStruct(adt_def, substs) => return ExprKind::Adt {
```
Comment (arielb1) [inline src/librustc_mir/hair/cx/expr.rs]: can't you have the same problem with unit variants (e.g. `Option::None`)?

Comment (nagisa) [inline src/librustc_mir/hair/cx/expr.rs]: Wow, I now feel kinda stupid not having noticed it myself 😃.

**Pull request merged**